src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs

```diff
@@ -119,7 +120,9 @@ private Block ProduceEmptyBlock(string payloadId, BlockHeader parentHeader, Payl
         private IBlockImprovementContext CreateBlockImprovementContext(string payloadId, BlockHeader parentHeader, PayloadAttributes payloadAttributes, Block currentBestBlock, DateTimeOffset startDateTime)
         {
             if (_logger.IsTrace) _logger.Trace($"Start improving block from payload {payloadId} with parent {parentHeader.ToString(BlockHeader.Format.FullHashAndNumber)}");
+            _currentBuildingContext?.Dispose();
```

Comment:
Can we have thread contention here? Probably not because engine API is itself using locks, but here we potentially could use Interlocked or a lock, for it to be independent.